freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeipaDownscaleNodeCalculatorService.java

```diff
@@ -0,0 +1,41 @@
+package com.sequenceiq.freeipa.service.stack;
+
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.springframework.stereotype.Service;
+
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.AvailabilityInfo;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.AvailabilityType;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.instance.InstanceMetadataType;
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.scale.DownscaleRequest;
+import com.sequenceiq.freeipa.entity.InstanceMetaData;
+import com.sequenceiq.freeipa.entity.Stack;
+
+@Service
+public class FreeipaDownscaleNodeCalculatorService {
+
+    public ArrayList<String> getDownscaleCandidates(Stack stack, AvailabilityInfo originalAvailabilityInfo, AvailabilityType targetAvailabilityType,
+            Set<String> instanceIdsToDelete) {
+        return instanceIdsToDelete.isEmpty()
+                ? calculateDownscaleCandidates(stack, originalAvailabilityInfo, targetAvailabilityType)
+                : new ArrayList<>(instanceIdsToDelete);
+    }
+
+    private ArrayList<String> calculateDownscaleCandidates(Stack stack, AvailabilityInfo originalAvailabilityInfo, AvailabilityType targetAvailabilityType) {
+        int instancesToRemove = originalAvailabilityInfo.getActualNodeCount() - targetAvailabilityType.getInstanceCount();
+        return stack.getNotDeletedInstanceMetaDataSet().stream()
+                .filter(imd -> imd.getInstanceMetadataType() != InstanceMetadataType.GATEWAY_PRIMARY)
+                .limit(instancesToRemove)
+                .map(InstanceMetaData::getInstanceId)
+                .collect(Collectors.toCollection(ArrayList::new));
+    }
+
+    public AvailabilityType getTargetAvailabilityType(DownscaleRequest request, int currentNodeCount) {
```

Comment:
it would be nice if you could replace `get` with something else, like `calculate` as there is a lof of get for this variable already